### ISSUE TITLE
Windows11 WAMP/XAMPP - Hello world works BUT Importing external module causes 500 Internal Error

### ISSUE THREAD

**Issue by gidzr** [GrahamDumpleton/mod_wsgi#862]
Hi Graham

I'm really excited to find your wsgi project. I'm hoping to use Python in LAMP and WAMP stacks without relying on exec/shell_exec commands.

Issue in a Nutshell
-----------------------------------------------
Everything setup successfully, with a successful Hello World. But when the my_function includes "import xxxx" from a pip installation, it causes a 500 internal error response. The system / wsgi is unable to find python modules installed with pip.
Precisely from Apache Error Log: \
```
[Sat Oct 28 11:33:59.020264 2023] [wsgi:error] [pid 31328:tid 2032] [client ::1:62110] ImportError: DLL load failed while importing _C: The specified module could not be found.\r, referer: h.t.t.p://localhost/www/project1/front_end.html
```
I have read Issues: https://github.com/GrahamDumpleton/mod_wsgi/issues/853 which is not relevant because the DLL does load, and https://github.com/GrahamDumpleton/mod_wsgi/issues/417 which is most similar as they have a module loading problem with Windows XAMPP. However, I have added all the paths correctly to the Windows environmental vars and there's no issue with users vs system access in how everything runs.

Pre-testing Non-WSGI approaches to confirm py scripts are good
-----------------------------------------------
1. Running py script from exec() / shell_exec() within WAMP: No errors, runs perfectly when calling py functions usnig imports
2. Running py script from Windows cli terminal commands: No errors, runs perfectly when calling py functions usnig imports.
3. I've also tested the scripts on remote Debian/Apache/Python host not using wsgi, and the python scripts have no issues. Definitely a WSGI issue.

**I haven't tested WSGI on LAMP.. only WAMP. I need to establish a mirror environment of prod on localhost for dev/test/build.**

Pertinent Questions
-----------------------------------------------
1. Is there a Windows environment variable that needs to be set for python modules?
2. Is wsgi having an issue in windows due to how python runs environments?
5. Is it a failure in how I've set up the httpd.conf file?
6. Is it simply a matter of adding a line in the httpd.conf file that points to the 'root' folder for modules eg. C:\Users\xxx\AppData\Local\Programs\Python\Python311\Lib\site-packages

Context and Setup
-----------------------------------------------
Machine: 
- Windows 11, i7 10 core, 16GRam

Stack: 
- Bitnami AMP stack via XAMPP, 
- Apache Server version: Apache/2.4.56 (Win64) Apache Lounge VS16 Server
- Python 3.11.4

Windows Environment Vars:
All Users / System
:: MOD_WSGI_APACHE_ROOTDIR: C:\xampp81\apache
:: Path: C:\Users\xxxx\AppData\Local\Programs\Python\Python311\Scripts\
:: Path: C:\Users\xxx\AppData\Local\Programs\Python\Python311\Lib\site-packages
:: Path: C:\Users\xxxx\AppData\Local\Programs\Python\Python311\

User xxxx
:: MOD_WSGI_APACHE_ROOTDIR: C:\xampp81\apache
:: Path: C:\Users\xxxx\AppData\Local\Programs\Python\Python311\Scripts\
:: Path: C:\Users\xxx\AppData\Local\Programs\Python\Python311\Lib\site-packages
:: Path: C:\Users\xxxx\AppData\Local\Programs\Python\Python311\

```
Apache httpd.conf
  DocumentRoot "C:/xampp81/htdocs"
  <Directory "C:/xampp81/htdocs">

    Options Indexes FollowSymLinks Includes MultiViews ExecCGI
      AllowOverride All
      Require all granted
      AddHandler wsgi-script .wsgi                #<----- added this for wsgi

  </Directory>

  #--> added the following based on wsgi instructions (mod_wsgi-express module-config)
  LoadFile "C:/Users/xxxx/AppData/Local/Programs/Python/Python311/python311.dll"
  LoadModule wsgi_module "C:/Users/xxxx/AppData/Local/Programs/Python/Python311/Lib/site-packages/mod_wsgi/server/mod_wsgi.cp311-win_amd64.pyd"
  WSGIPythonHome "C:/Users/xxxx/AppData/Local/Programs/Python/Python311"
```

Testing Setup 1 - Basic point to point (installation test)
-----------------------------------------------
1. Created a server script: pyserve.wsgi
```
from wsgiref.util import setup_testing_defaults
def application(environ, start_response):
   status = '200 OK'
   output = ('Hello World!').encode()
   response_headers = [('Content-type', 'text/plain'),
               ('Content-Length', str(len(output)))]
   start_response(status, response_headers)
   return [output]`
```

7. In front_end.html, created an AJAX call to the pyserve.wsgi server
```
<script>
(()=>{
  $.ajax({
        type: 'POST',  // Or 'GET', depending on what your application expects
        url: '../python/pyserve.wsgi',
        success: function(response) {
           console.log('success')
           console.log(response);
        },
        error: function(error) {
           console.log(error);
        }
     });
})()
</script>
```

### Testing Setup 1 Results
Success! Hello World prints to the browser console.log

Testing Setup 2 - Calling a function
-----------------------------------------------
1. Created a server script: pyserve.wsgi
```
from wsgiref.util import setup_testing_defaults
import sys; 
sys.path.append('www/project1/python/');
from my_script import my_function

def application(environ, start_response):
   status = '200 OK'
   output = my_function().encode()
   response_headers = [('Content-type', 'text/plain'),
               ('Content-Length', str(len(output)))]
   start_response(status, response_headers)
   return [output]
```

3. In my_script.py, basic hello world test

```
#!/usr/bin/env python

def my_function():
    return "hello world"
```

### Testing Setup 2 - Results
Success! Hello World prints to the browser console.log

Testing Setup 3 - Calling a function which imports modules
-----------------------------------------------
SAME AS Testing Setup 2, but modifying my_script.py to the following:

```
#!/usr/bin/env python

import chromadb         #<----------- added this line

def my_function():
    return "hello world"
```

### Testing Setup 3 - Results
FAILS.  Console Error:

```
500 Internal Server Error
Internal Server Error
The server encountered an internal error or
misconfiguration and was unable to complete
your request. Please contact the server administrator
Apache/2.4.56 (Win64) OpenSSL/1.1.1t PHP/8.1.17 mod_wsgi/4.9.4 Python/3.11 Server at localhost Port 80
```

### Testing Setup 3 - Apache Error Log
```
[Sat Oct 28 11:33:59.020264 2023] [wsgi:error] [pid 31328:tid 2032] [client ::1:62110] ImportError: DLL load failed while importing _C: The specified module could not be found.\r, referer: http://localhost/www/project1/front_end.html
```

**Comment by GrahamDumpleton** [comment]
Are you using a Python virtual environment?

Or are you installing Python packages into per user site packages directory?

What arguments are you supplying to `pip` when installing packages?

**Comment by gidzr** [comment]
Hi Graham, fast response!

Still doing testing and tried these after reading your documentation: https://modwsgi.readthedocs.io/en/master/configuration-directives/WSGIPythonPath.html
**httpd.conf**
WSGIPythonPath "C:/Users/xxxx/AppData/Local/Programs/Python/Python311/Lib/site-packages"
**Window Env Vars**
User and System: PYTHONPATH: "C:/Users/xxxx/AppData/Local/Programs/Python/Python311/Lib/site-packages"

In answer to your questions:
1. Are you installing Python packages into per user site packages directory?
I'm using root for everything, and installed with cmd terminal in Admin mode. But where I can, I'm adding variables to user and system level to eliminate any blind spots. I'm not using venv/anaconda.. I just wanted a straight one-time installation single environment to avoid environmental issues.

2. What arguments are you supplying to pip when installing packages?
From windows cmd terminal
$ pip install chromadb

I repeat this using cmd terminal in Admin and normal mode to confirms all installation requirements already met (ie. universally can find the site package), and I run python and pip in both Admin or normal mode to ensure the Windows env vars are set and can find these commands/applications from any location or admin mode.

I'm convinced (maybe wrongly) that it's a path issue.. so I'm trying different combinations of things.

So far with the above settings, it's doesn't return internal 500 error, it just hangs with "Pending" in browser console.log.

**Comment by GrahamDumpleton** [comment]
If `_C` part of `chromadb`? Where does the DLL for that reside?

**Comment by gidzr** [comment]
I think you are referring to the _C directory
eg. torch does. ie. C:\Users\xxx\AppData\Local\Programs\Python\Python311\Lib\site-packages\torch\_C

chromadb doesn't install a _C folder

**Comment by GrahamDumpleton** [comment]
Then what Python C extensions are in that directory, and what DLLs do they need to work?

Python on Windows can sometime be confusing about it error messages.

**Comment by gidzr** [comment]
.pyc files are in all the directories

I'm a little confused, as the python scripts run from windows terminal cmd: "$ python my_script.py" works, and using php exec('python my_script.py') also works. 

I thought the mod_wsgi would enable apache to serve python files effectively the same way the php_fpm / mod_php enable php files to be served. With the addition of a .wsgi file to act as a wrapper for headers.

What have I messed up?

**Comment by GrahamDumpleton** [comment]
You haven't necessarily mucked up anything. Have seen cases in the past where because of how Python C extension modules were compiled on Windows they would not find associated DLL files when rather than running Python command line, you are using Python embedded in a different application (eg., Apache httpd). The errors from failed DLL loading are often not very helpful and is hard to work out what the problem is.

Also, because Apache generally runs as a distinct Window service user, it doesn't have the same runtime environment as your user where you installed stuff. So if such DLLs are usually only found because their location was add to your `PATH`, then they will not be found by Apache httpd service.

I am not a Windows user I really don't know how to debug these sorts of problems. All I know is that in some extreme cases in the past it was necessary to use the `LoadFile` directive in Apache to manually load the DLL into Apache prior to the Python C extension needing it being imported.

Windows just isn't a very good platform for using Python embedded in other applications as some support for embedding Python on Windows is broken and hacks are required in mod_wsgi to even get it mostly working.

Best I can suggest is try and narrow it down to the specific Python module/extension which is failing import, and then work out what DLL or C extension it is using so can work out why it cannot be found.

In other words, start out with trying to import `torch._C` and see if that works. If not, look in `__init__.py` of that package if isn't a C extension but a directory and see what modules it in turn imports. Try importing each of those directly instead, and narrow it down so can work out which specific one is failing.

**Comment by gidzr** [comment]
LOL.. that's a shame.

Appreciate your help in trying to narrow down a testing approach. That certainly helps.

Just to paraphrase to see if I get it.

1. Using PATH windows environment variables may not work, even if I add path to the user AND system, because there could be third 'apache' user type that isn't shown (eg. like a binami, xampp, or daemon user) which would also need the PATH variable.

2. Loadfile directive in the httpd.conf could be used to manually load all the modules.

3. I could try "import torch._C" in the script to see if that works, or drilling down, adding path to sys.path.amend('') to drill down location of module for the script to see if this approach can get it loaded.

... 

If my understanding is correct per above, could you clarify, 
re 1) is it possible to add an environment variable for apache using httpd.conf or setting env variable with windows cli terminal the same way it's done for linux? (export varname=value)?

re 2) if I wanted to manually add location of modules to the httpd.conf file, would I need to add:
LoadFile "C:/Users/xxxx/AppData/Local/Programs/Python/Python311/site-package/packagename/..../module.pyc" ??
or is there a folder approach I could use?
or would I be looking for a DLL file?

**Comment by GrahamDumpleton** [comment]
For (1), there is no way to set environment variable for PATH in Apache configuration such that it affects the current process loading of DLLs. Presumably, if PATH needs to be modified it needs to be done in the registry for the service user Apache runs as. Whether this is true and not sure as am not a windows user and that is based on what little Windows knowledge I have from 10-20 years ago.

For (2), you aren't preloading the Python extension (pyc) files. LoadFile is only if need to force load a Windows DLL file.

For (3), there is no need to modify sys.path. You are just aiming to narrow down which module/extension is failing to load. So progressively importing more specific module/extension by dotted Python module import path.

**Comment by gidzr** [comment]
Hi Graham

I have some good news, and some follow-up questions.

Good news everyone
------------------------------
**in httpd.conf** added the following:
```
LoadFile "C:/Users/xxxx/AppData/Local/Programs/Python/Python311/python311.dll"
LoadModule wsgi_module "C:/Users/xxxx/AppData/Local/Programs/Python/Python311/Lib/site-packages/mod_wsgi/server/mod_wsgi.cp311-win_amd64.pyd"
WSGIPythonHome "C:/Users/xxxx/AppData/Local/Programs/Python/Python311"

WSGIPythonPath "C:/Users/xxxx/AppData/Local/Programs/Python/Python311/Scripts; C:/Users/xxxx/AppData/Local/Programs/Python/Python311/Lib/site-packages"   ///<--- newly added these two settings for python path
```

**WINDOWS environment vars, in both the User and System settings** added the following
```
PATH: "C:\Users\gidzv\AppData\Local\Programs\Python\Python311\; C:\Users\gidzv\AppData\Local\Programs\Python\Python311\Lib\site-packages; C:\Users\gidzv\AppData\Local\Programs\Python\Python311\Scripts"
```
AND newly added this  //<------- newly added these
```
PYTHONPATH:  "C:\Users\gidzv\AppData\Local\Programs\Python\Python311\Lib\site-packages; C:\Users\gidzv\AppData\Local\Programs\Python\Python311\Scripts"
```

Then fully updated my_script.py away from print(hello world) to imported module function:

```
from pypdf import PdfReader

def my_function():
   reader = PdfReader(/folder/pathto/file.pdf')
   page_text=[];
   for page in reader.pages:
      page_text = page.extract_text()

   return page_text
```

which SUCCESSFULLY returned the pdf contents to the console.log (yayy)

Follow-up questions
------------------------------
1. Error catching and parsing to browser console.log
2. Restarting apache every time the my_script.py is changed

Re 1 (error catching): Whenever there's an error in the python script, rather than outputting the actual problem (eg. wrong file path to the .pdf I'm reading with pypdf), I simply receive a 500 internal server error, not an output of the full error. Am I supposed to be running a try()-catch(err) in the .wsgi file to enable this? I can't find any error handling on output of py scripts in the documentation.

Re 2 (restarting): Is the only way to propagate changes to my_script to reset apache every single time? This adds quite a bit of wasted time to the coding-troubleshooting time. Is there a config setting that enables apache to autocheck the script each time it loads? is this : WSGILazyInitialization? If so, where does it go (not working in httpd.config)

**Comment by GrahamDumpleton** [comment]
For (1), uncaught Python exceptions should be logged to the Apache error log.

Note that if using certain Python WSGI frameworks you may have to configure them to log details of errors to the logs as otherwise will not log them.

For (2), your options on Windows are limited because of how Apache works. See:

* https://modwsgi.readthedocs.io/en/master/user-guides/reloading-source-code.html

Some of that config you have shouldn't usually be required, but then for whatever reason embedded Python doesn't work as it is meant to for some people. I don't know if is because how Python was installed, per user, or for whole machine, but if you have it working I will not suggest you fiddle with it.

**Comment by GrahamDumpleton** [comment]
BTW, rather than build WSGI applications from scratch, suggesting using Flask framework.

https://flask.palletsprojects.com/en/3.0.x/

Just remember to configure it to log errors.

**Comment by gidzr** [comment]
Hi Graham

Premature celebration.. I thought I had it all worked out, but I was caught out by the requirement to keep restarting apache for every alteration of the my_script.py. When I started apache, I was back to square 1 re import chromadb.

However, I was successful in running the script using the module pypdf, which uses "from xxx import xxx" pattern.

I think I'm stuck with a number of issues to resolve to make mod_wsgi a practical solution for me at this point. :(
a) need to have changes from my_script.py propagate without having to restart apache
b) need to have python errors funnelled to the console.log, as currently any error in the script will result in 500 internal error
c) need to figure out how to send arguments from AJAX to .wsgi and then to my_function(args)

Good idea re flask. For now I'll probably stick with the exec(), shell_exec(), and revisit as tech debt when the app is no longer a prototype.
